smart-contracts/contracts/cl-vault/src/vault/swap.rs

```diff
@@ -84,20 +99,129 @@ pub fn swap(
+                dex_router_address.to_string(),
+                &QueryMsg::BestPathForPair {
+                    offer_asset: offer_asset.into(),
+                    ask_asset: ask_asset.into(),
+                    exclude_paths: None,
+                    offer_amount: token_in_amount,
+                },
+            )?;
+            let best_out = match best_path.clone() {
+                Some(best_path) => best_path.return_amount,
+                None => 0u128.into(),
+            };
+
+            // if we need to force the route
+            if force_swap_route {
+                match recommended_swap_route {
+                    Some(recommended_swap_route) => execute_swap_operations(
+                        dex_router_address,
+                        recommended_swap_route,
+                        token_out_min_amount,
+                        token_in_denom,
+                        token_in_amount,
+                    ),
+                    None => Err(ContractError::TryForceRouteWithoutRecommendedSwapRoute {}),
+                }
+            } else if best_out.is_zero() && recommended_out.is_zero() {
+                Ok(swap_exact_amount_in(
+                    &env,
+                    pool_route,
+                    token_in_amount,
+                    token_in_denom,
+                    token_out_min_amount,
+                ))
+            } else if (best_out.ge(&recommended_out)) {
+                execute_swap_operations(
+                    dex_router_address,
+                    best_path.unwrap().operations.into(),
+                    token_out_min_amount,
+                    token_in_denom,
+                    token_in_amount,
+                )
+            } else {
+                // recommended_out > best_out
+                execute_swap_operations(
+                    dex_router_address,
+                    recommended_swap_route.unwrap(), // will be some here
+                    token_out_min_amount,
+                    token_in_denom,
+                    token_in_amount,
+                )
+            }
         }
-        .into();
+        None => Ok(swap_exact_amount_in(
+            &env,
+            pool_route,
+            token_in_amount,
+            token_in_denom,
+            token_out_min_amount,
+        )),
+    };
+
+    Ok(swap_msg?)
+}
+
+fn swap_exact_amount_in(
+    env: &Env,
+    pool_route: SwapAmountInRoute,
+    token_in_amount: Uint128,
+    token_in_denom: &String,
+    token_out_min_amount: Uint128,
+) -> CosmosMsg {
+    osmosis_std::types::osmosis::poolmanager::v1beta1::MsgSwapExactAmountIn {
+        sender: env.contract.address.to_string(),
+        routes: vec![pool_route],
+        token_in: Some(OsmoCoin {
+            denom: token_in_denom.to_string(),
+            amount: token_in_amount.to_string(),
+        }),
+        token_out_min_amount: token_out_min_amount.to_string(),
+    }
+    .into()
+}
+
+fn execute_swap_operations(
```

Comment:
this would be better to be prefixed as `get_`, this would be more clear as the function returns a CosmosMsg, also above.

Reply:
I've thought about that. Minor but lmk if you want me to dig into the correct strategy here